pre-processing.py

```diff
@@ -0,0 +1,228 @@
+    "year",
+    "state",
+    "state_tax_revenue",
+    "state_taxable_income",
+]
+
+spmu = person.groupby(SPMU_COLUMNS, observed=False)[PERSON_COLUMNS].sum().reset_index()
+spmu[["fica", "fedtaxac", "stataxac"]] *= -1
+spmu.rename(columns={"person": "numper"}, inplace=True)
+
+# write pre-processed dfs to csv files
+person.to_csv("person.csv.gz", compression="gzip")
+spmu.to_csv("spmu.csv.gz", compression="gzip")
+
+# create boolean column for individual's poverty status, 1=poor
+person["poor"] = person.spmthresh > person.spmtotres
+
+# create a column for all selected demographic variables
+# that will be used to calculate poverty rates
+demog_cols = [
+    "person",
+    "adult",
+    "child",
+    "black",
+    "white_non_hispanic",
+    "hispanic",
+    "pwd",
+    "non_citizen",
+    "non_citizen_adult",
+    "non_citizen_child",
+]
+
+poor_pop = person.loc[person.poor]
+
+# calculate weighted sum of people living in poverty
+mdf.weighted_sum(poor_pop, demog_cols, "asecwt")
+# calculate poverty RATE for each DEMOGRAPHIC in US
+pov_rate_us = mdf.weighted_sum(poor_pop, demog_cols, "asecwt") / mdf.weighted_sum(
+    person, demog_cols, w="asecwt"
+)
+# add name to series
+pov_rate_us.name = "US"
+# calculate poverty RATE for each group by state
+pov_rates = mdf.weighted_sum(
+    poor_pop, demog_cols, "asecwt", groupby="state"
+) / mdf.weighted_sum(person, demog_cols, w="asecwt", groupby="state")
+
+# append US statistics as additional 'state'
+pov_df = pov_rates.append(pov_rate_us)
+
+# melt df from wide to long format
+pov_df = pov_df.melt(ignore_index=False, var_name="demog")
+# insert column indicating metric in question
+pov_df.insert(loc=1, column="metric", value="pov_rate")
+
+
+##
+# calculate POPULATION for each DEMOGRAPHIC in US
+pop_us = mdf.weighted_sum(person, demog_cols, w="asecwt")
+# add name to series
+pop_us.name = "US"
+# calculate POPULATION for each group by state
+pop_states = mdf.weighted_sum(person, demog_cols, w="asecwt", groupby="state")
+# append US statistics as additional 'state'
+pop_df = pop_states.append(pop_us)
+# melt df from wide to long format
+pop_df = pop_df.melt(ignore_index=False, var_name="demog")
+pop_df.insert(loc=1, column="metric", value="pop")
+
+# concat poverty and population dfs
+demog_stats = pd.concat([pov_df, pop_df])
+# write to csv file
+demog_stats.to_csv("demog_stats.csv.gz", compression="gzip")
+
+# Caluclate original gini
+person["spm_resources_per_person"] = person.spmtotres / person.numper
+# Caluclate original gini for US
+gini_us = pd.Series(mdf.gini(df=person, col="spm_resources_per_person", w="asecwt"))
+# add name to series
+gini_us.index = ["US"]
+
+
+# calculate gini for each group by state
+gini_states = mdf.gini(
+    df=person, col="spm_resources_per_person", w="asecwt", groupby="state"
+)
+
+# append US statistics as additional 'state'
+gini_ser = gini_states.append(gini_us)
+gini_ser.name = "gini"
+
+# Calculate the original poverty gap
+spmu["poverty_gap"] = np.where(
```

Comment:
Might not want to since you're slicing by state and `mdf.poverty_gap` doesn't yet have a `groupby` arg, but FYI you can replace this thru 200 with:
```
poverty_gap_us = pd.Series(mdf.poverty_gap(spmu, "spmtotres", "spmthresh", "spmwt"), index=["US"])
```

Reply:
Oh that's dope. Can you apply it to a `pandas`  `groupby` ?

Reply:
Could with a lambda, e.g. if `states` is a `pd.Series` with each state, I think this would work:
```
poverty_gap_state = states.apply(lambda x: mdf.poverty_gap(spmu[spmu.state == x], "spmtotres", "spmthresh", "spmwt"))
```